File: src/main/java/net/internetisalie/lunar/lang/lexer/LuaTokenTypes.java
/*
 * Copyright 2010 Jon S Akhtar (Sylvanaar)
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package net.internetisalie.lunar.lang.lexer;

import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import net.internetisalie.lunar.lang.psi.LuaElementType;

/**
 * Interface that contains all tokens returned by LuaLexer
 *
 * @author sylvanaar
 */
public interface LuaTokenTypes {
    //IFileElementType FILE = new IFileElementType(Language.findInstance(LuaLanguage.class));
    /**
     * Wrong token. Use for debugger needs
     */
    IElementType WRONG = TokenType.BAD_CHARACTER;


    /* **************************************************************************************************
     *  Whitespaces & NewLines
     * ****************************************************************************************************/

    IElementType NL_BEFORE_LONGSTRING = new LuaElementType("newline after longstring start bracket");
    IElementType WS = TokenType.WHITE_SPACE;
    IElementType NEWLINE = new LuaElementType("new line");

    /* **************************************************************************************************
     *  Comments
     * ****************************************************************************************************/

    IElementType MARKER = new LuaElementType("line label marker");

    IElementType SHEBANG = new LuaElementType("shebang - should ignore");

    IElementType LONGCOMMENT = new LuaElementType("long comment");
    IElementType SHORTCOMMENT = new LuaElementType("short comment");

    IElementType LONGCOMMENT_BEGIN = new LuaElementType("long comment start bracket");
    IElementType LONGCOMMENT_END = new LuaElementType("long comment end bracket");

    /* **************************************************************************************************
     *  Identifiers
     * ****************************************************************************************************/

    IElementType IDENTIFIER = new LuaElementType("identifier");

    /* **************************************************************************************************
     *  Integers & floats
     * ****************************************************************************************************/

    IElementType NUMBER = new LuaElementType("number");

    /* **************************************************************************************************
     *  Strings & regular expressions
     * ****************************************************************************************************/

    IElementType STRING = new LuaElementType("string");
    IElementType LONGSTRING = new LuaElementType("long string");

    IElementType LONGSTRING_BEGIN = new LuaElementType("long string start bracket");
    IElementType LONGSTRING_END = new LuaElementType("long string end bracket");

    /* **************************************************************************************************
     *  Common tokens: operators, braces etc.
     * ****************************************************************************************************/

    IElementType DIV = new LuaElementType("/");
    IElementType MULT = new LuaElementType("*");
    IElementType LPAREN = new LuaElementType("(");
    IElementType RPAREN = new LuaElementType(")");
    IElementType LBRACK = new LuaElementType("[");
    IElementType RBRACK = new LuaElementType("]");
    IElementType LCURLY = new LuaElementType("{");
    IElementType RCURLY = new LuaElementType("}");
    IElementType COLON = new LuaElementType(":");
    IElementType COMMA = new LuaElementType(",");
    IElementType DOT = new LuaElementType(".");
    IElementType ASSIGN = new LuaElementType("=");
    IElementType SEMI = new LuaElementType(";");
    IElementType EQ = new LuaElementType("==");
    IElementType NE = new LuaElementType("~=");
    IElementType PLUS = new LuaElementType("+");
    IElementType MINUS = new LuaElementType("-");
    IElementType GE = new LuaElementType(">=");
    IElementType GT = new LuaElementType(">");
    IElementType EXP = new LuaElementType("^");
    IElementType LE = new LuaElementType("<=");
    IElementType LT = new LuaElementType("<");
    IElementType ELLIPSIS = new LuaElementType("...");
    IElementType CONCAT = new LuaElementType("..");
    IElementType GETN = new LuaElementType("#");
    IElementType MOD = new LuaElementType("%");

    // New Operators
    IElementType INTDIV = new LuaElementType("//");
    IElementType AMP = new LuaElementType("&");
    IElementType NEG = new LuaElementType("~");
    IElementType PIPE = new LuaElementType("|");
    IElementType BSR = new LuaElementType(">>");
    IElementType BSL = new LuaElementType("<<");

    /* **************************************************************************************************
     *  Keywords
     * ****************************************************************************************************/

    IElementType IF = new LuaElementType("if");
    IElementType ELSE = new LuaElementType("else");
    IElementType ELSEIF = new LuaElementType("elseif");
    IElementType WHILE = new LuaElementType("while");
    IElementType WITH = new LuaElementType("with");

    IElementType THEN = new LuaElementType("then");
    IElementType FOR = new LuaElementType("for");
    IElementType IN = new LuaElementType("in");
    IElementType RETURN = new LuaElementType("return");
    IElementType BREAK = new LuaElementType("break");

    IElementType CONTINUE = new LuaElementType("continue");
    IElementType TRUE = new LuaElementType("true");
    IElementType FALSE = new LuaElementType("false");
    IElementType NIL = new LuaElementType("nil");
    IElementType FUNCTION = new LuaElementType("function");

    IElementType DO = new LuaElementType("do");
    IElementType NOT = new LuaElementType("not");
    IElementType AND = new LuaElementType("and");
    IElementType OR = new LuaElementType("or");
    IElementType LOCAL = new LuaElementType("local");

    IElementType REPEAT = new LuaElementType("repeat");
    IElementType UNTIL = new LuaElementType("until");
    IElementType END = new LuaElementType("end");

    IElementType GOTO = new LuaElementType("goto");
}

File: src/main/kotlin/net/internetisalie/lunar/lang/lexer/LuaTokenTypes.kt
/*
 * Copyright 2010 Jon S Akhtar (Sylvanaar)
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package net.internetisalie.lunar.lang.lexer

import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType
import net.internetisalie.lunar.lang.psi.LuaElementType

/**
 * Holder of all tokens returned by LuaLexer.
 *
 * @author sylvanaar
 */
object LuaTokenTypes {
    /**
     * Wrong token. Use for debugger needs
     */
    @JvmField val WRONG: IElementType = TokenType.BAD_CHARACTER

    // Whitespaces & NewLines

    @JvmField val NL_BEFORE_LONGSTRING: IElementType = LuaElementType("newline after longstring start bracket")
    @JvmField val WS: IElementType = TokenType.WHITE_SPACE
    @JvmField val NEWLINE: IElementType = LuaElementType("new line")

    // Comments

    @JvmField val MARKER: IElementType = LuaElementType("line label marker")

    @JvmField val SHEBANG: IElementType = LuaElementType("shebang - should ignore")

    @JvmField val LONGCOMMENT: IElementType = LuaElementType("long comment")
    @JvmField val SHORTCOMMENT: IElementType = LuaElementType("short comment")

    @JvmField val LONGCOMMENT_BEGIN: IElementType = LuaElementType("long comment start bracket")
    @JvmField val LONGCOMMENT_END: IElementType = LuaElementType("long comment end bracket")

    // Identifiers

    @JvmField val IDENTIFIER: IElementType = LuaElementType("identifier")

    // Integers & floats

    @JvmField val NUMBER: IElementType = LuaElementType("number")

    // Strings & regular expressions

    @JvmField val STRING: IElementType = LuaElementType("string")
    @JvmField val LONGSTRING: IElementType = LuaElementType("long string")

    @JvmField val LONGSTRING_BEGIN: IElementType = LuaElementType("long string start bracket")
    @JvmField val LONGSTRING_END: IElementType = LuaElementType("long string end bracket")

    // Common tokens: operators, braces etc.

    @JvmField val DIV: IElementType = LuaElementType("/")
    @JvmField val MULT: IElementType = LuaElementType("*")
    @JvmField val LPAREN: IElementType = LuaElementType("(")
    @JvmField val RPAREN: IElementType = LuaElementType(")")
    @JvmField val LBRACK: IElementType = LuaElementType("[")
    @JvmField val RBRACK: IElementType = LuaElementType("]")
    @JvmField val LCURLY: IElementType = LuaElementType("{")
    @JvmField val RCURLY: IElementType = LuaElementType("}")
    @JvmField val COLON: IElementType = LuaElementType(":")
    @JvmField val COMMA: IElementType = LuaElementType(",")
    @JvmField val DOT: IElementType = LuaElementType(".")
    @JvmField val ASSIGN: IElementType = LuaElementType("=")
    @JvmField val SEMI: IElementType = LuaElementType(";")
    @JvmField val EQ: IElementType = LuaElementType("==")
    @JvmField val NE: IElementType = LuaElementType("~=")
    @JvmField val PLUS: IElementType = LuaElementType("+")
    @JvmField val MINUS: IElementType = LuaElementType("-")
    @JvmField val GE: IElementType = LuaElementType(">=")
    @JvmField val GT: IElementType = LuaElementType(">")
    @JvmField val EXP: IElementType = LuaElementType("^")
    @JvmField val LE: IElementType = LuaElementType("<=")
    @JvmField val LT: IElementType = LuaElementType("<")
    @JvmField val ELLIPSIS: IElementType = LuaElementType("...")
    @JvmField val CONCAT: IElementType = LuaElementType("..")
    @JvmField val GETN: IElementType = LuaElementType("#")
    @JvmField val MOD: IElementType = LuaElementType("%")

    // New Operators
    @JvmField val INTDIV: IElementType = LuaElementType("//")
    @JvmField val AMP: IElementType = LuaElementType("&")
    @JvmField val NEG: IElementType = LuaElementType("~")
    @JvmField val PIPE: IElementType = LuaElementType("|")
    @JvmField val BSR: IElementType = LuaElementType(">>")
    @JvmField val BSL: IElementType = LuaElementType("<<")

    // Keywords

    @JvmField val IF: IElementType = LuaElementType("if")
    @JvmField val ELSE: IElementType = LuaElementType("else")
    @JvmField val ELSEIF: IElementType = LuaElementType("elseif")
    @JvmField val WHILE: IElementType = LuaElementType("while")
    @JvmField val WITH: IElementType = LuaElementType("with")

    @JvmField val THEN: IElementType = LuaElementType("then")
    @JvmField val FOR: IElementType = LuaElementType("for")
    @JvmField val IN: IElementType = LuaElementType("in")
    @JvmField val RETURN: IElementType = LuaElementType("return")
    @JvmField val BREAK: IElementType = LuaElementType("break")

    @JvmField val CONTINUE: IElementType = LuaElementType("continue")
    @JvmField val TRUE: IElementType = LuaElementType("true")
    @JvmField val FALSE: IElementType = LuaElementType("false")
    @JvmField val NIL: IElementType = LuaElementType("nil")
    @JvmField val FUNCTION: IElementType = LuaElementType("function")

    @JvmField val DO: IElementType = LuaElementType("do")
    @JvmField val NOT: IElementType = LuaElementType("not")
    @JvmField val AND: IElementType = LuaElementType("and")
    @JvmField val OR: IElementType = LuaElementType("or")
    @JvmField val LOCAL: IElementType = LuaElementType("local")

    @JvmField val REPEAT: IElementType = LuaElementType("repeat")
    @JvmField val UNTIL: IElementType = LuaElementType("until")
    @JvmField val END: IElementType = LuaElementType("end")

    @JvmField val GOTO: IElementType = LuaElementType("goto")
    @JvmField val GLOBAL: IElementType = LuaElementType("global")
}
